nixos/modules/services/networking/crab-hole.nix

```
@@ -0,0 +1,230 @@
+
+  settingsFormat = pkgs.formats.toml { };
+in
+{
+  options = {
+    services.crab-hole = {
+      enable = lib.mkEnableOption "Crab-hole Service";
+
+      package = lib.mkPackageOption pkgs "crab-hole" { };
+
+      supplementaryGroups = lib.mkOption {
+        type = lib.types.listOf lib.types.str;
+        default = [ ];
+        example = [ "acme" ];
+        description = "Adds additional groups to the crab-hole service. Can be useful to prevent permission issues.";
+      };
+
+      settings = lib.mkOption {
+        description = "Crab-holes config. See big example https://github.com/LuckyTurtleDev/crab-hole/blob/main/example-config.toml";
+
+        example = {
+          downstream = [
+            {
+              listen = "localhost";
+              port = 8080;
+              protocol = "udp";
+            }
+            {
+              certificate = "dns.example.com.crt";
+              dns_hostname = "dns.example.com";
+              key = "dns.example.com.key";
+              listen = "[::]";
+              port = 8055;
+              protocol = "https";
+              timeout_ms = 3000;
+            }
+          ];
+          api = {
+            admin_key = "1234";
+            listen = "127.0.0.1";
+            port = 8080;
+            show_doc = true;
+          };
+          blocklist = {
+            allow_list = [
+              "file:///allowed.txt"
+            ];
+            include_subdomains = true;
+            lists = [
+              "https://raw.githubusercontent.com/StevenBlack/hosts/master/alternates/fakenews-gambling-porn/hosts"
+              "https://s3.amazonaws.com/lists.disconnect.me/simple_tracking.txt"
+              "file:///blocked.txt"
+            ];
+          };
+          upstream = {
+            name_servers = [
+              {
+                protocol = "tls";
+                socket_addr = "[2606:4700:4700::1111]:853";
+                tls_dns_name = "1dot1dot1dot1.cloudflare-dns.com";
+                trust_nx_responses = false;
+              }
+              {
+                protocol = "tls";
+                socket_addr = "1.1.1.1:853";
+                tls_dns_name = "1dot1dot1dot1.cloudflare-dns.com";
+                trust_nx_responses = false;
+              }
+            ];
+            options = {
+              validate = false;
+            };
+          };
+        };
+
+        type = lib.types.submodule {
+          freeformType = settingsFormat.type;
+          options = {
+            blocklist = {
+              include_subdomains = lib.mkEnableOption "Include Subdomains";
+              lists = lib.mkOption {
+                type = lib.types.listOf (lib.types.either lib.types.str lib.types.path);
+                default = [ ];
+                description = "List of blocklists. If files are added via url, make sure the service has access to them!";
+                apply = x: (lib.lists.forEach x (v: if builtins.isPath v then ("file://${v}") else v));
```

If you have a repeated type like this then it's probably a good idea to make it DRY

(Also, using `builtins.map` here would be better since you can do an eta-reduction)

```suggestion
                apply = map (v: if builtins.isPath v then "file://${v}" else v);
```